src/cargo/core/compiler/custom_build.rs

```diff
@@ -536,6 +540,44 @@ impl BuildOutput {
     }
 }
 
+fn prepare_metabuild<'a, 'cfg>(
+    cx: &Context<'a, 'cfg>,
+    unit: &Unit<'a>,
+    deps: &[String],
+) -> CargoResult<()> {
+    let mut output = Vec::new();
+    let available_deps = cx.dep_targets(unit);
+    // Filter out optional dependencies, and look up the actual lib name.
+    let meta_deps: Vec<_> = deps
+        .iter()
+        .filter_map(|name| {
+            available_deps
+                .iter()
+                .find(|u| u.pkg.name().as_str() == name.as_str())
+                .map(|dep| dep.target.crate_name())
+        })
+        .collect();
+    for dep in &meta_deps {
+        output.push(format!("extern crate {};\n", dep));
+    }
+    output.push("fn main() {\n".to_string());
+    for dep in &meta_deps {
+        output.push(format!("    {}::metabuild();\n", dep));
+    }
+    output.push("}".to_string());
+    let output = output.join("");
+    let path = cx.files().metabuild_path(unit);
+    let changed = if let Ok(orig) = fs::read_to_string(&path) {
+        orig != output
+    } else {
+        true
+    };
+    if changed {
+        fs::write(&path, output)?;
+    }
```

Comment:
For clarity and potential future reuse, please extract this logic into a "write if changed" function.